### ISSUE TITLE
In oculus go controller, if we press button_A and button_Enter together the app is crashing.The cause for this is geavrf library is throwing IllegalArguementException.

### ISSUE THREAD

**Issue by talk2saroj** [Samsung/GearVRf#2000]

**Comment by talk2saroj** [comment]
How to fix this issue.What is the current version for org.gearvrf:framework

**Comment by talk2saroj** [comment]
I am using 4.0.1-Snapshot

**Comment by liaxim** [comment]
Can you share the stack trace? Thanks.

**Comment by liaxim** [comment]
The snapshots were somewhat outdated (last uploaded mid-Sep). At the end of Sep we made a change to support better controller multiple key presses. I just updated the snapshots. Please try again.

**Comment by talk2saroj** [comment]
> Can you share the stack trace? Thanks.
Hi,please find the stack trace here.
[gearvrf_logs.txt](https://github.com/Samsung/GearVRf/files/2481445/gearvrf_logs.txt)

**Comment by talk2saroj** [comment]
Hi ,thank you very much for quick response.
I tried refreshing the dependencies and got the following error and its making sense.Since, the new library has updated the abstract method's signature,and I am using the previous signature.

### error: <anonymous com.example.org.gvrfapplication.HomeScreenActivity$2> i
s not abstract and does not override abstract method onControllerEvent(CONTROLLER_KEYS[],Vector3f,Quaternionf,PointF,boolean,Vector3f,Vector3f) in IActivityEvents
        {
        ^
E:\SureLock_VR\GVRFApplication\app\src\main\java\com\example\org\gvrfapplication\HomeScreenActivity.java:122: error: method does not override or implement a method from a supertype
            @Override
            ^

 But still inside the code, i am getting the previous abstract method to override, but not the new one.
Am i missing something.Please update.

**Comment by talk2saroj** [comment]
The following  is the signature for previous abstract method...

 **void onControllerEvent(GVRGearCursorController.CONTROLLER_KEYS key, Vector3f position, Quaternionf orientation, PointF touchpadPoint, boolean touched, Vector3f angularAcceleration,
                           Vector3f angularVelocity);**

**Comment by liaxim** [comment]
Something must be cached somewhere. Try deleting manually the gradle cache - <your-user-dir>/.gradle/caches/.

**Comment by talk2saroj** [comment]
Hi,after deleting the cache,ut started working.But on press of any
button,no action is happening.
How to handle this,please let me knkw.
Thank you

**Comment by liaxim** [comment]
@talk2saroj Even single key presses do not work? For any key? Unfortunately I don't have an Oculus Go handy so I am afraid there will be a lot of back and forth.

**Comment by talk2saroj** [comment]
No,Button_A press is not working.But Button_Enter is working.On double tap
of Button_Enter, it s performing click operation.Inside the app, Back
button is not working. Home button is working fine.

**Comment by talk2saroj** [comment]
For time being, can you please share me the library which you released at the end of Sep with better controller multiple key presses.
Thanks.

**Comment by talk2saroj** [comment]
Button_A is Button.PrimaryIndexTrigger and Button_Enter is
Button.PrimaryTouchPad.These names come from the gearvrf library.

**Comment by liaxim** [comment]
Ok.. So you are saying that the onControllerEvent callback never gets the keycodes for Button_A or Button_Back? Does it get any other keycode instead?

**Comment by talk2saroj** [comment]
This is the callback for controller buttons.
 

**@Override
            public void onControllerEvent(GVRGearCursorController.CONTROLLER_KEYS[] keys, Vector3f position, Quaternionf orientation, PointF touchpadPoint, boolean touched, Vector3f angularAcceleration, Vector3f angularVelocity)
            {
                if(keys!=null)
                {
                    Logger.logInfo("Keys pressed are:"+keys);
                 
                }
            }**

When I am pressing Button_A, the callback is coming but in the keys,it is saying "All elements are null".
And when I press Back Button, the callback is coming and the key contains the Button_back , but, activity is not going back.
And for Button_Enter,the callback is coming and the key contains Button_Enter,but the click is happening on double tap,not on single tap.

**Comment by liaxim** [comment]
Button_A has a different key code on Oculus Go. Easy enough to remap that. The back key is a special case: while the key is mapped correctly it is not routed the same as on a phone. Will need to come up with a way to detect and handle Oculus Go specifically.

Button_Enter works for me as expected when it comes to tapping it. Is there any of our demo apps that show the double tap problem you've described? Thanks.

**Comment by talk2saroj** [comment]
you can try with this project-GearVRf-Demos-master\gvr-videoplayer .Actually,I am not able to run this project after updating to the library you shared because of some errors.But, in my project, I am facing the issue of double tap with this library.

**Comment by liaxim** [comment]
Ok, what I am seeing is that the Android events we generate to feed to the Android views (which gvr-videoplayer heavily uses) are not quite right hence the double-clicks or no clicks behavior.

**Comment by talk2saroj** [comment]
I came across some other vr app by Samsung named as Samsung VR with package name "com.samsung.vrvideo" which is working fine.
All the touch and click events are working properly.Do you have any idea which library they might be using. 
If you have some other version of library where these issues are fixed,Please share with me.
Thanks.

**Comment by liaxim** [comment]
We are looking into the issue and will have a fix in few days. Samsung VR is using something else. If you insist I can give you the commit id off which to build the framework for yourself. But it is going to be crashing on multiple key presses. You could patch that up.

**Comment by talk2saroj** [comment]
Ok,I can wait for the fix.
Thank you very much for the continuous support.

**Comment by liaxim** [comment]
@talk2saroj Back and trigger key handling should be fixed. Uploaded new snapshots. Double-tapping under investigation still.

**Comment by liaxim** [comment]
@talk2saroj It would be great if you could provide feedback and maybe open a new issue in case you experience some problem.
